server2/Arcadia.Assistant/Arcadia.Assistant.Sharepoint/SharepointSynchronizator.cs

```diff
@@ -0,0 +1,171 @@
+    using System.Threading;
+    using System.Threading.Tasks;
+
+    using CalendarEvent;
+
+    using Employees.Contracts;
+
+    using ExternalStorages.Abstractions;
+
+    using Microsoft.Extensions.Logging;
+
+    using SickLeaves.Contracts;
+
+    using Vacations.Contracts;
+
+    using WorkHoursCredit.Contracts;
+
+    public sealed class SharepointSynchronizator
+    {
+        #region ctor
+
+        public SharepointSynchronizator(ISickLeaves sickLeaves, IVacations vacations, IWorkHoursCredit workouts, ISharepointDepartmentsCalendarsSettings departmentsCalendarsSettings, ILogger? logger)
+        {
+            this.sickLeaves = sickLeaves;
+            this.vacations = vacations;
+            this.workouts = workouts;
+            this.departmentsCalendarsSettings = departmentsCalendarsSettings;
+            this.logger = logger;
+        }
+
+        #endregion
+
+        #region publci interface
+
+        public async Task Synchronize(IEmployees employees, IEnumerable<string> departments, IExternalStorage storage, CancellationToken cancellationToken)
+        {
+            var vacationsSync = new EmployeeVacationsSynchronization(storage, this.logger);
+            var workoutsSync = new EmployeeWorkoutsSynchronization(storage, this.logger);
+            var sickLeavesSync = new EmployeeSickLeavesSynchronization(storage, this.logger);
+
+            var sharepointCalendars = departments.Distinct().ToDictionary(x => x,
+                dId => this.GetSharepointCalendarsByDepartment(dId).ToDictionary(x => x,
+                    async cal => await this.GetAllSharepointItemsForCalendar(storage, cal)));
+
+            foreach (var departmentId in departments)
+            {
+                var departmentEmployes = await employees.FindEmployeesAsync(EmployeesQuery.Create().ForDepartment(departmentId), cancellationToken);
+                var employeeIds = departmentEmployes.Select(x => x.EmployeeId).ToArray();
+
+                var employeeVacations = await this.vacations.GetCalendarEventsByEmployeeAsync(employeeIds, cancellationToken);
+                var employeeVacationValues = employeeVacations.Values.SelectMany(x => x).ToDictionary(x => CspCalendarEventIdParser.GetCalendarEventIdFromCspId(x.VacationId, CalendarEventTypes.Vacation), x => x);
+
+                var employeeWorkouts = await this.workouts.GetCalendarEventsByEmployeeMapAsync(employeeIds, cancellationToken);
+                var employeeWorkoutsValues = employeeWorkouts.Values.SelectMany(x => x).ToDictionary(x => CspCalendarEventIdParser.GetCalendarEventIdFromCspId(x.ChangeId, CalendarEventTypes.Workout), x => x);
+
+                var employeeSickLeaves = await this.sickLeaves.GetCalendarEventsByEmployeeMapAsync(employeeIds, cancellationToken);
+                var employeeSickLeavesValues = employeeSickLeaves.Values.SelectMany(x => x).ToDictionary(x => CspCalendarEventIdParser.GetCalendarEventIdFromCspId(x.SickLeaveId, CalendarEventTypes.Sickleave), x => x);
```

Comment:
These three too. This will reduce waiting time.

Reply:
Done